File: c/tests/02_basic_parsing.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>

#include "test.hpp"

bool cb_ran = false;
string out;

int
callback_help(CheckArg *ca, const char *, const char *) {
  cb_ran = true;
  unique_ptr<char> xout(checkarg_autohelp(ca));
  out = xout.get();
  return CA_ALLOK;
}


TEST_CASE("parsing: help using --help", "[parsing]") {
  vector<const char *> argv = {
    "/test01",
  };

  argv.push_back(GENERATE("--help", "-h"));

  cb_ran = false;
  out.clear();

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);

  // the default callback calls exit(0), which is bad here
  // checkarg_add_autohelp();

  // it seems one can also not just pass a lambda to a C function
  /* checkarg_add_cb(ca.get(), */
  /*   'h', "help", */
  /*   [&out,&cb_ran](CheckArg *ca, const char*, const char*) -> int { */
  /*     cb_ran = true; */
  /*     unique_ptr<char> xout(checkarg_autohelp(ca)); */
  /*     out = xout.get(); */
  /*     return CA_ALLOK; */
  /*   }, */
  /*   "show this help message and exit" */
  /* ); */
  checkarg_add_cb(
    ca.get(), 'h', "help", &callback_help, "show this help message and exit", 0, NULL);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);
  auto ret = checkarg_parse(ca.get(), argv.size(), (char **)argv.data());

  CHECK(ret == CA_ALLOK);  // -h ran fine
  CHECK(cb_ran);           // callback actually ran
  REQUIRE(out ==
    "Usage: test01 [options]\n"
    "\n"
    "Options:\n"
//    "   -h, --help         show this help message and exit\n"
//    "   -i, --input=INPUT  file to read from\n"
    "   -h, --help   show this help message and exit\n"
    "   -i, --input  file to read from\n"
  ); // callback had access to CA object
}

TEST_CASE("parsing: correct option and value", "[parsing]") {
  const vector<const char *> argv = {
    "/usr/bin/name",
    "-i",
    "input.in",
  };

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), argv.size(), (char **)argv.data());

  // value ok
  REQUIRE(ret == CA_ALLOK);
  // option detected
  REQUIRE(checkarg_isset(ca.get(), "input"));
  // value stored correctly
  REQUIRE(checkarg_value(ca.get(), "input") == string("input.in"));
}


TEST_CASE("parsing: invalid option", "[parsing]") {
  const vector<const char *> argv = {
    "/usr/bin/name",
    "-x",
  };

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), argv.size(), (char **)argv.data());

  REQUIRE(ret == CA_INVOPT);  // invalid option
}

TEST_CASE("parsing: missing value", "[parsing]") {
  const vector<const char *> argv = {
    "/usr/bin/name",
    "-i",
  };

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), argv.size(), (char **)argv.data());

  REQUIRE(ret == CA_MISSVAL);  // missing value
}


TEST_CASE("parsing: empty argv (std::vector)", "[parsing]") {
  const vector<const char *> argv{};

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), argv.size(), (char **)argv.data());

  REQUIRE(ret == CA_ERROR);  // empty argv
}

TEST_CASE("parsing: argv == NULL (argc=0, argv=NULL)", "[parsing]") {
  const char **argv = NULL;

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), 0, (char **)argv);

  REQUIRE(ret == CA_ERROR);  // null argv
}

TEST_CASE("parsing: empty argv (argc=0, argv={NULL})", "[parsing]") {
  const char **argv = {NULL};

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), 0, (char **)argv);

  REQUIRE(ret == CA_ERROR);  // empty argv
}

TEST_CASE("parsing: empty argv (argc=1, argv={NULL})", "[parsing]") {
  const char **argv = {NULL};

  CheckArgUPtr ca(checkarg_new("test01", NULL, NULL), &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  auto ret = checkarg_parse(ca.get(), 1, (char **)argv);

  REQUIRE(ret == CA_ERROR);  // empty argv
}

File: rust/tests/02_basic_parsing.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>
#![feature(assert_matches)]

#[cfg(test)]
mod tests {
  use checkarg::{CheckArg, ValueType, RC};
  use std::assert_matches::assert_matches;

  fn triggering_help(option: &str) {
    let argv = vec!["/test01", option];

    let mut cb_ran = 0;
    let mut out = String::new();
    {
      let mut ca = CheckArg::new("test01");
      // the default callback calls exit(0), which is bad here
      // ca.add_autohelp();
      ca.add_cb(
        'h',
        "help",
        "show this help message and exit",
        |cx, _opt, _val| {
          cb_ran += 1;
          out = cx.autohelp();
          Ok(())
        },
        ValueType::None,
        None,
      );
      ca.add(
        'i',
        "input",
        "file to read from",
        ValueType::Required,
        Some(""),
      );

      let rc = ca.parse(&argv);

      assert_matches!(rc, RC::Ok, "parsing failed");
    }
    assert_eq!(cb_ran, 1, "callback did not run");

    assert_eq!(
      out,
      "\
Usage: test01 [options]

Options:
   -h, --help   show this help message and exit
   -i, --input  file to read from\n"
    );
  }

  #[test]
  fn triggering_help_long_option() { triggering_help("--help") }

  #[test]
  fn triggering_help_short_option() { triggering_help("-h") }

  fn setup_simple_ca() -> CheckArg<'static> {
    let mut ca = CheckArg::new("name");
    ca.add(
      'i',
      "input",
      "file to read from",
      ValueType::Required,
      Some(""),
    );
    ca
  }

  #[test]
  fn correct_option_and_value() {
    let mut ca = setup_simple_ca();

    let argv = vec!["/usr/bin/name", "-i", "input.in"];

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::Ok, "parsing failed");
    assert!(ca.isset("input"));
    assert_matches!(ca.value("input"), Some(input) if input == "input.in");
    assert_eq!(ca.value("input").unwrap(), "input.in");
  }

  #[test]
  fn invalid_option() {
    let mut ca = setup_simple_ca();

    let argv = vec!["/usr/bin/name", "-x"];

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::InvOpt, "parse should detect invalid option");
  }

  #[test]
  fn missing_value() {
    let mut ca = setup_simple_ca();

    let argv = vec!["/usr/bin/name", "-i"];

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::MissVal, "parse should detect missing value");
  }

  #[test]
  #[should_panic(expected = "argv must have at least one element")]
  fn empty_argv_vector() {
    let mut ca = setup_simple_ca();

    let argv = vec![];

    ca.parse(&argv);
  }
}
